src/endpoints/minidump.rs

```diff
@@ -162,3 +162,109 @@ pub fn configure(app: ServiceApp) -> ServiceApp {
         r.method(Method::POST).with(handle_minidump_request);
     })
 }
+
+#[cfg(test)]
+mod tests {
+    use std::fs;
+
+    use actix_web::test::TestServer;
+    use reqwest::{multipart, Client, StatusCode};
+    use tokio::runtime::Runtime;
+
+    use crate::app::ServiceState;
+    use crate::config::Config;
+    use crate::test;
+    use crate::types::SymbolicationResponse;
+
+    #[tokio::test]
+    async fn test_basic() {
+        test::setup();
+
+        let service = ServiceState::create(Config::default()).unwrap();
+        let server = TestServer::with_factory(move || crate::server::create_app(service.clone()));
+
+        let file_contents = fs::read("tests/fixtures/windows.dmp").unwrap();
+        let file_part = multipart::Part::bytes(file_contents).file_name("windows.dmp");
+
+        let form = multipart::Form::new()
+            .part("upload_file_minidump", file_part)
+            .text("sources", "[]");
+
+        let response = Client::new()
+            .post(&server.url("/minidump"))
+            .multipart(form)
+            .send()
+            .await
+            .unwrap();
+
+        assert_eq!(response.status(), StatusCode::OK);
+
+        let body = response.text().await.unwrap();
+        let response = serde_json::from_str::<SymbolicationResponse>(&body).unwrap();
+        insta::assert_yaml_snapshot!(response);
+    }
+
+    #[test]
+    fn test_integration_microsoft() {
+        // TODO: Move this test to E2E tests
+        test::setup();
+
+        // XXX: This cannot be a `#[tokio::test]` because the service spawns a tokio runtime
+        // internally. Dropping this within another runtime created by `tokio::test` is not allowed.
+        let service = ServiceState::create(Config::default()).unwrap();
+
+        let s = service.clone();
+        let server = TestServer::with_factory(move || crate::server::create_app(s.clone()));
+        let source = test::microsoft_symsrv();
+
+        Runtime::new().unwrap().block_on(async {
+            let file_contents = fs::read("tests/fixtures/windows.dmp").unwrap();
+            let file_part = multipart::Part::bytes(file_contents).file_name("windows.dmp");
+
+            let form = multipart::Form::new()
+                .part("upload_file_minidump", file_part)
+                .text("sources", serde_json::to_string(&vec![source]).unwrap())
+                .text("options", r#"{"dif_candidates":true}"#);
+
+            let response = Client::new()
+                .post(&server.url("/minidump"))
+                .multipart(form)
+                .send()
+                .await
+                .unwrap();
+
+            assert_eq!(response.status(), StatusCode::OK);
+
+            let body = response.text().await.unwrap();
+            let response = serde_json::from_str::<SymbolicationResponse>(&body).unwrap();
+            insta::assert_yaml_snapshot!(response);
+        });
+
+        drop(service); // ensure the service outlives the runtime.
```

Comment:
I suspect this comment isn't true.  I think needs to be dropped here because waiting on implicit drop makes it race with something else.  Maybe with something the test framework does to the other threads spawned by this thread or so?  I actually wonder when exactly the runtime is torn down, wonder if you tried assigning it and explicitly dropping it in various orders.  Obviously this is a massive pain because it only deadlocks on CI, so I don't mind that we don't know.

Reply:
Could be that dropping the `ServiceState` explicitly shuts down the procspawn pool?

Reply:
There is no difference between implicit and explicit drop. My best guess is that by dropping here we ensure that there is no lexical drop inside of the runtime. We could replace this statement with just `service;` and it would still work.